File: HanSeokhyeon/Lifeboat.java
/*
구명보트

무게제한을 넘지 않게 태워야하고
최소한으로 배를 사용해서 구하고 싶다.
 */

import java.util.Arrays;

public class Lifeboat {
    public static void main(String[] args) {
        Lifeboat lifeboat = new Lifeboat();
        int solution = lifeboat.solution(new int[]{70, 50, 80, 50}, 100);
        System.out.println("solution = " + solution);
        solution = lifeboat.solution(new int[]{70, 80, 50}, 100);
        System.out.println("solution = " + solution);
    }

    public int solution(int[] people, int limit) {
        int answer = 0;
        int[] weights = Arrays.stream(people).sorted().toArray();
        int start = 0; int end = people.length-1;
        while (start <= end) {
            int weightSum = weights[end--]; // 제일 무거운거 하나 싣고
            while (weightSum + weights[start] <= limit) { // 더이상 안들어갈 때 까지
                weightSum += weights[start++]; // 가벼운거 계속 싣기
            }
            answer++;
        }
        return answer;
    }
}

File: MinJunKim/lifeboat.js

// 젤 작은거 젤 큰거부터 따져서 한도 초과면 젤큰거 뺴고 진행, 답은 1증가
// 젤큰거 작은거가 한도 안넘으면 두개 뺴고 다음것들 진행, 답 1증가
// 개수 한개 남으면 답 1증가
function solution(people, limit) {
  var answer = 0;
  let peopleOrder = people.sort((a,b) => a-b);
  let i = 0;
  let j = people.length - 1;
  while (i < j) {
    if (peopleOrder[i] + peopleOrder[j] > limit) {
      j--;
    } else {
      i++;
      j--;
      
    }
      answer++;
  }
  if (i === j) {
    answer++;
  }
  return answer;
}
